src/hypertable.c

```diff
@@ -3014,3 +3020,112 @@ ts_hypertable_update_dimension_partitions(const Hypertable *ht)
 
 	return false;
 }
+
+/*
+ * hypertable_osm_range_update
+ * 0 hypertable REGCLASS,
+ * 1 range_start,
+ * 2 range_end,
+ * 3 empty=false
+ * If empty is set to true then the range will be set to invalid range
+ * but the overlap flag will be unset, indicating that no data is managed
+ * by OSM and therefore timescaledb optimizations can be applied.
+ */
+TS_FUNCTION_INFO_V1(ts_hypertable_osm_range_update);
+Datum
+ts_hypertable_osm_range_update(PG_FUNCTION_ARGS)
+{
+	Oid relid = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	Hypertable *ht;
+	const Dimension *time_dim;
+	Cache *hcache;
+
+	Oid time_type; /* required for resolving the argument types, should match the hypertable
+					  partitioning column type */
+
+	hcache = ts_hypertable_cache_pin();
+	ht = ts_resolve_hypertable_from_table_or_cagg(hcache, relid, true);
+	Assert(ht != NULL);
+	time_dim = hyperspace_get_open_dimension(ht->space, 0);
+
+	Ensure(time_dim != NULL,
+		   "could not find time dimension for hypertable %s.%s",
+		   quote_identifier(NameStr(ht->fd.schema_name)),
+		   quote_identifier(NameStr(ht->fd.table_name)));
+
+	time_type = ts_dimension_get_partition_type(time_dim);
+
+	int32 osm_chunk_id = ts_chunk_get_osm_chunk_id(ht->fd.id);
+	if (osm_chunk_id == INVALID_CHUNK_ID)
+		elog(ERROR,
+			 "no OSM chunk found for hypertable %s.%s",
+			 quote_identifier(NameStr(ht->fd.schema_name)),
+			 quote_identifier(NameStr(ht->fd.table_name)));
+
+	int32 dimension_slice_id = ts_chunk_get_osm_slice_id(osm_chunk_id, time_dim->fd.id);
+
+	/*
+	 * range_start, range_end arguments must be converted to internal representation
+	 * a NULL start value is interpreted as INT64_MAX - 1 and a NULL end value is
+	 * interpreted as infinity.
+	 * Passing both start and end NULL values will reset the range to the default range an
+	 * OSM chunk is given upon creation, which is [INT64_MAX - 1, INT64_MAX]
+	 */
+	int64 range_start_internal, range_end_internal;
+	if (PG_ARGISNULL(1))
+		range_start_internal = PG_INT64_MAX - 1;
+	else
+		range_start_internal = ts_time_value_to_internal(PG_GETARG_DATUM(1), time_type);
+	if (PG_ARGISNULL(2))
+		range_end_internal = PG_INT64_MAX;
+	else
+		range_end_internal = ts_time_value_to_internal(PG_GETARG_DATUM(2), time_type);
+
+	if (range_start_internal > range_end_internal)
+		ereport(ERROR, errmsg("dimension slice range_end cannot be less than range_start"));
+
+	bool osm_chunk_empty = PG_GETARG_BOOL(3);
+
+	bool overlap = false, range_invalid = false;
+
+	ScanTupLock tuplock = {
+		.lockmode = LockTupleExclusive,
+		.waitpolicy = LockWaitBlock,
+	};
+	DimensionSlice *slice =
+		ts_dimension_slice_scan_by_id_and_lock(dimension_slice_id, &tuplock, CurrentMemoryContext);
+
+	if (!slice)
+		ereport(ERROR, errmsg("could not find slice with id %d", dimension_slice_id));
+	overlap = ts_osm_chunk_range_overlaps(dimension_slice_id,
+										  slice->fd.dimension_id,
+										  range_start_internal,
+										  range_end_internal);
```

Comment:
I think we should treat this as error now, since the range of osm chunk can never overlap with other chunks since it has to be contiguous so no non-osm chunk can exist in that range.

Reply:
So we’d like OSM to call the update_range function and set the range to invalid whenever there is a hole (hole could mean either chunk has not been created in tsdb, or chunk exists in tsdb but has not been tiered). Perhaps instead of error, it would be better to print a NOTICE or WARNING here that the range overlaps and it will be set to invalid range instead. What do you think?

Reply:
I am still puzzled by why we need this. Going to discuss this in more detail in slack.

Reply:
After discussion: We need non overlapping ranges  for chunks- otherwise tuple routing breaks.This not easy to fix as part of this PR. So the current expectation is that OSM chunk saves valid range information in the timescaledb catalog only if ranges are contiguous.